src/steps/build-team-to-member-mapping/index.ts

```diff
@@ -29,34 +29,99 @@ const step: IntegrationStep<HerokuIntegrationConfig> = {
     },
   ],
   dependsOn: [TEAM_STEP, MEMBER_STEP],
-  async executionHandler({ instance, jobState }) {
-    const heroku = new HerokuClient(instance.config);
+  async executionHandler(context) {
+    const { instance, jobState, logger } = context;
 
+    const heroku = new HerokuClient(instance.config);
     const userIdMap = await createUserIdMap(jobState);
+    const teamEntities = await collectTeamEntities(jobState);
+
+    logger.info(
+      {
+        numTeamEntities: teamEntities.length,
+      },
+      'Total teams to fetch members for',
+    );
+
+    for (const teamEntity of teamEntities) {
+      try {
+        const teamMembers = await heroku.getTeamMembers(
+          teamEntity.id as string,
+        );
+
+        logger.info(
+          {
+            numTeamMembers: teamMembers.length,
+          },
+          'Successfully fetched team members',
+        );
 
-    await jobState.iterateEntities({ _type: TEAM_TYPE }, async (team) => {
-      const teamMembers = await heroku.getTeamMembers(team.id as string);
-
-      for (const teamMember of teamMembers) {
-        const userId = teamMember.user ? teamMember.user.id : null;
-        const user = userIdMap.get(userId);
-
-        if (user) {
-          await jobState.addRelationships([
-            createDirectRelationship({
-              _class: RelationshipClass.HAS,
-              from: team,
-              to: user,
-              properties: { role: teamMember.role },
-            }),
-          ]);
+        await createTeamHasUserRelationships({
+          userIdMap,
+          teamMembers,
+          teamEntity: teamEntity,
+          jobState,
+        });
+      } catch (err) {
+        if (err.code === 'PROVIDER_AUTHORIZATION_ERROR') {
+          logger.publishEvent({
+            name: 'missing_scope',
+            description: `Could not fetch team members. Missing required OAuth scope (endpoint=${err.endpoint}, scope=global)`,
+          });
+
+          return;
         }
+
+        throw err;
       }
-    });
+    }
   },
 };
 
-export default step;
+async function collectTeamEntities(jobState: JobState): Promise<Entity[]> {
+  const teamEntities: Entity[] = [];
+
+  await jobState.iterateEntities(
```

Comment:
It would be nice to have some way of early exiting this function in the future.

Perhaps something similar to Attempt:

```typescript
await jobState.iterateEntities({
  _type: 'the_type',
  onError(err, ctx) {
    ctx.abort();
  }
}, async (e) {
  ...
});
```

Reply:
I agree, there needs to be something like this. I'm curious whether the `ctx.abort()` pattern is more idiomatic than having the callback return some kind of signal?